// @angular/language-service/src/types.js
"use strict";
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
Object.defineProperty(exports, "__esModule", { value: true });
var compiler_cli_1 = require("@angular/compiler-cli");
exports.BuiltinType = compiler_cli_1.BuiltinType;
/**
 * The kind of diagnostic message.
 *
 * @experimental
 */
var DiagnosticKind;
(function (DiagnosticKind) {
    DiagnosticKind[DiagnosticKind["Error"] = 0] = "Error";
    DiagnosticKind[DiagnosticKind["Warning"] = 1] = "Warning";
})(DiagnosticKind = exports.DiagnosticKind || (exports.DiagnosticKind = {}));
;
var RefactoringType = (function () {
    function RefactoringType() {
    }
    return RefactoringType;
}());
RefactoringType.SYMBOL_RENAME = "SYMBOL_RENAME";
RefactoringType.RESOURCE_RENAME = "RESOURCE_RENAME";
RefactoringType.RESOURCE_MOVE = "RESOURCE_MOVE";
exports.RefactoringType = RefactoringType;
var RefactoringParameter = (function () {
    function RefactoringParameter() {
    }
    return RefactoringParameter;
}());
RefactoringParameter.DESTINATION = "destination";
RefactoringParameter.UPDATE_REFERENCES = "updateReferences";
RefactoringParameter.UPDATE_EXPORTED_SYMBOLS = "updateExportedSymbols";
exports.RefactoringParameter = RefactoringParameter;
