// @angular/compiler/src/summary_resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var injectable_1 = require("./injectable");
var SummaryResolver = (function () {
    function SummaryResolver() {
    }
    return SummaryResolver;
}());
exports.SummaryResolver = SummaryResolver;
var JitSummaryResolver = (function () {
    function JitSummaryResolver() {
        this._summaries = new Map();
    }
    JitSummaryResolver.prototype.isLibraryFile = function (fileName) { return false; };
    JitSummaryResolver.prototype.getLibraryFileName = function (fileName) { return null; };
    JitSummaryResolver.prototype.resolveSummary = function (reference) {
        return this._summaries.get(reference) || null;
    };
    JitSummaryResolver.prototype.getSymbolsOf = function (filePath) { return []; };
    JitSummaryResolver.prototype.getImportAs = function (reference) { return reference; };
    JitSummaryResolver.prototype.addSummary = function (summary) { this._summaries.set(summary.symbol, summary); };
    return JitSummaryResolver;
}());
JitSummaryResolver.decorators = [
    { type: injectable_1.CompilerInjectable },
];
/** @nocollapse */
JitSummaryResolver.ctorParameters = function () { return []; };
exports.JitSummaryResolver = JitSummaryResolver;
