// @angular/core/src/animation/animation_transition.js
"use strict";
var animation_transition_event_1 = require('./animation_transition_event');
var AnimationTransition = (function () {
    function AnimationTransition(_player, _fromState, _toState, _totalTime) {
        this._player = _player;
        this._fromState = _fromState;
        this._toState = _toState;
        this._totalTime = _totalTime;
    }
    AnimationTransition.prototype._createEvent = function (phaseName) {
        return new animation_transition_event_1.AnimationTransitionEvent({
            fromState: this._fromState,
            toState: this._toState,
            totalTime: this._totalTime,
            phaseName: phaseName
        });
    };
    AnimationTransition.prototype.onStart = function (callback) {
        var _this = this;
        var fn = Zone.current.wrap(function () { return callback(_this._createEvent('start')); }, 'player.onStart');
        this._player.onStart(fn);
    };
    AnimationTransition.prototype.onDone = function (callback) {
        var _this = this;
        var fn = Zone.current.wrap(function () { return callback(_this._createEvent('done')); }, 'player.onDone');
        this._player.onDone(fn);
    };
    return AnimationTransition;
}());
exports.AnimationTransition = AnimationTransition;
