// @angular/compiler/src/private_import_core.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
"use strict";
var core_1 = require('@angular/core');
exports.isDefaultChangeDetectionStrategy = core_1.__core_private__.isDefaultChangeDetectionStrategy;
exports.ChangeDetectorStatus = core_1.__core_private__.ChangeDetectorStatus;
exports.LifecycleHooks = core_1.__core_private__.LifecycleHooks;
exports.LIFECYCLE_HOOKS_VALUES = core_1.__core_private__.LIFECYCLE_HOOKS_VALUES;
exports.ReflectorReader = core_1.__core_private__.ReflectorReader;
exports.ViewContainer = core_1.__core_private__.ViewContainer;
exports.CodegenComponentFactoryResolver = core_1.__core_private__.CodegenComponentFactoryResolver;
exports.ComponentRef_ = core_1.__core_private__.ComponentRef_;
exports.AppView = core_1.__core_private__.AppView;
exports.DebugAppView = core_1.__core_private__.DebugAppView;
exports.NgModuleInjector = core_1.__core_private__.NgModuleInjector;
exports.registerModuleFactory = core_1.__core_private__.registerModuleFactory;
exports.ViewType = core_1.__core_private__.ViewType;
exports.view_utils = core_1.__core_private__.view_utils;
exports.DebugContext = core_1.__core_private__.DebugContext;
exports.StaticNodeDebugInfo = core_1.__core_private__.StaticNodeDebugInfo;
exports.devModeEqual = core_1.__core_private__.devModeEqual;
exports.UNINITIALIZED = core_1.__core_private__.UNINITIALIZED;
exports.ValueUnwrapper = core_1.__core_private__.ValueUnwrapper;
exports.TemplateRef_ = core_1.__core_private__.TemplateRef_;
exports.RenderDebugInfo = core_1.__core_private__.RenderDebugInfo;
exports.Console = core_1.__core_private__.Console;
exports.reflector = core_1.__core_private__.reflector;
exports.Reflector = core_1.__core_private__.Reflector;
exports.ReflectionCapabilities = core_1.__core_private__.ReflectionCapabilities;
exports.NoOpAnimationPlayer = core_1.__core_private__.NoOpAnimationPlayer;
exports.AnimationPlayer = core_1.__core_private__.AnimationPlayer;
exports.AnimationSequencePlayer = core_1.__core_private__.AnimationSequencePlayer;
exports.AnimationGroupPlayer = core_1.__core_private__.AnimationGroupPlayer;
exports.AnimationKeyframe = core_1.__core_private__.AnimationKeyframe;
exports.AnimationStyles = core_1.__core_private__.AnimationStyles;
exports.ANY_STATE = core_1.__core_private__.ANY_STATE;
exports.DEFAULT_STATE = core_1.__core_private__.DEFAULT_STATE;
exports.EMPTY_STATE = core_1.__core_private__.EMPTY_STATE;
exports.FILL_STYLE_FLAG = core_1.__core_private__.FILL_STYLE_FLAG;
exports.prepareFinalAnimationStyles = core_1.__core_private__.prepareFinalAnimationStyles;
exports.balanceAnimationKeyframes = core_1.__core_private__.balanceAnimationKeyframes;
exports.clearStyles = core_1.__core_private__.clearStyles;
exports.collectAndResolveStyles = core_1.__core_private__.collectAndResolveStyles;
exports.renderStyles = core_1.__core_private__.renderStyles;
exports.ViewMetadata = core_1.__core_private__.ViewMetadata;
exports.ComponentStillLoadingError = core_1.__core_private__.ComponentStillLoadingError;
exports.AnimationTransition = core_1.__core_private__.AnimationTransition;
