// @angular/core/src/core_private_export.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
"use strict";
var animation_constants_1 = require('./animation/animation_constants');
var animation_group_player_1 = require('./animation/animation_group_player');
var animation_keyframe_1 = require('./animation/animation_keyframe');
var animation_player_1 = require('./animation/animation_player');
var animation_sequence_player_1 = require('./animation/animation_sequence_player');
var animationUtils = require('./animation/animation_style_util');
var animation_styles_1 = require('./animation/animation_styles');
var animation_transition_1 = require('./animation/animation_transition');
var application_tokens = require('./application_tokens');
var change_detection_util = require('./change_detection/change_detection_util');
var constants = require('./change_detection/constants');
var console = require('./console');
var debug = require('./debug/debug_renderer');
var reflective_provider = require('./di/reflective_provider');
var compiler_1 = require('./linker/compiler');
var component_factory = require('./linker/component_factory');
var component_factory_resolver = require('./linker/component_factory_resolver');
var debug_context = require('./linker/debug_context');
var ng_module_factory = require('./linker/ng_module_factory');
var ng_module_factory_loader = require('./linker/ng_module_factory_loader');
var template_ref = require('./linker/template_ref');
var view = require('./linker/view');
var view_container = require('./linker/view_container');
var view_type = require('./linker/view_type');
var view_utils = require('./linker/view_utils');
var lifecycle_hooks = require('./metadata/lifecycle_hooks');
var metadata_view = require('./metadata/view');
var reflection = require('./reflection/reflection');
var reflection_capabilities = require('./reflection/reflection_capabilities');
var reflector_reader = require('./reflection/reflector_reader');
var api = require('./render/api');
var decorators = require('./util/decorators');
var lang_1 = require('./util/lang');
exports.__core_private__ = {
    isDefaultChangeDetectionStrategy: constants.isDefaultChangeDetectionStrategy,
    ChangeDetectorStatus: constants.ChangeDetectorStatus,
    constructDependencies: reflective_provider.constructDependencies,
    LifecycleHooks: lifecycle_hooks.LifecycleHooks,
    LIFECYCLE_HOOKS_VALUES: lifecycle_hooks.LIFECYCLE_HOOKS_VALUES,
    ReflectorReader: reflector_reader.ReflectorReader,
    CodegenComponentFactoryResolver: component_factory_resolver.CodegenComponentFactoryResolver,
    ComponentRef_: component_factory.ComponentRef_,
    ViewContainer: view_container.ViewContainer,
    AppView: view.AppView,
    DebugAppView: view.DebugAppView,
    NgModuleInjector: ng_module_factory.NgModuleInjector,
    registerModuleFactory: ng_module_factory_loader.registerModuleFactory,
    ViewType: view_type.ViewType,
    view_utils: view_utils,
    ViewMetadata: metadata_view.ViewMetadata,
    DebugContext: debug_context.DebugContext,
    StaticNodeDebugInfo: debug_context.StaticNodeDebugInfo,
    devModeEqual: change_detection_util.devModeEqual,
    UNINITIALIZED: change_detection_util.UNINITIALIZED,
    ValueUnwrapper: change_detection_util.ValueUnwrapper,
    RenderDebugInfo: api.RenderDebugInfo,
    TemplateRef_: template_ref.TemplateRef_,
    ReflectionCapabilities: reflection_capabilities.ReflectionCapabilities,
    makeDecorator: decorators.makeDecorator,
    DebugDomRootRenderer: debug.DebugDomRootRenderer,
    Console: console.Console,
    reflector: reflection.reflector,
    Reflector: reflection.Reflector,
    NoOpAnimationPlayer: animation_player_1.NoOpAnimationPlayer,
    AnimationPlayer: animation_player_1.AnimationPlayer,
    AnimationSequencePlayer: animation_sequence_player_1.AnimationSequencePlayer,
    AnimationGroupPlayer: animation_group_player_1.AnimationGroupPlayer,
    AnimationKeyframe: animation_keyframe_1.AnimationKeyframe,
    prepareFinalAnimationStyles: animationUtils.prepareFinalAnimationStyles,
    balanceAnimationKeyframes: animationUtils.balanceAnimationKeyframes,
    flattenStyles: animationUtils.flattenStyles,
    clearStyles: animationUtils.clearStyles,
    renderStyles: animationUtils.renderStyles,
    collectAndResolveStyles: animationUtils.collectAndResolveStyles,
    APP_ID_RANDOM_PROVIDER: application_tokens.APP_ID_RANDOM_PROVIDER,
    AnimationStyles: animation_styles_1.AnimationStyles,
    ANY_STATE: animation_constants_1.ANY_STATE,
    DEFAULT_STATE: animation_constants_1.DEFAULT_STATE,
    EMPTY_STATE: animation_constants_1.EMPTY_STATE,
    FILL_STYLE_FLAG: animation_constants_1.FILL_STYLE_FLAG,
    ComponentStillLoadingError: compiler_1.ComponentStillLoadingError,
    isPromise: lang_1.isPromise,
    isObservable: lang_1.isObservable,
    AnimationTransition: animation_transition_1.AnimationTransition
};
