// @angular/core/src/core.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
/**
 * @module
 * @description
 * Entry point from which you should import all public core APIs.
 */
__export(require('./metadata'));
__export(require('./version'));
__export(require('./util'));
__export(require('./di'));
var application_ref_1 = require('./application_ref');
exports.createPlatform = application_ref_1.createPlatform;
exports.assertPlatform = application_ref_1.assertPlatform;
exports.destroyPlatform = application_ref_1.destroyPlatform;
exports.getPlatform = application_ref_1.getPlatform;
exports.PlatformRef = application_ref_1.PlatformRef;
exports.ApplicationRef = application_ref_1.ApplicationRef;
exports.enableProdMode = application_ref_1.enableProdMode;
exports.isDevMode = application_ref_1.isDevMode;
exports.createPlatformFactory = application_ref_1.createPlatformFactory;
exports.NgProbeToken = application_ref_1.NgProbeToken;
var application_tokens_1 = require('./application_tokens');
exports.APP_ID = application_tokens_1.APP_ID;
exports.PACKAGE_ROOT_URL = application_tokens_1.PACKAGE_ROOT_URL;
exports.PLATFORM_INITIALIZER = application_tokens_1.PLATFORM_INITIALIZER;
exports.APP_BOOTSTRAP_LISTENER = application_tokens_1.APP_BOOTSTRAP_LISTENER;
var application_init_1 = require('./application_init');
exports.APP_INITIALIZER = application_init_1.APP_INITIALIZER;
exports.ApplicationInitStatus = application_init_1.ApplicationInitStatus;
__export(require('./zone'));
__export(require('./render'));
__export(require('./linker'));
var debug_node_1 = require('./debug/debug_node');
exports.DebugElement = debug_node_1.DebugElement;
exports.DebugNode = debug_node_1.DebugNode;
exports.asNativeElements = debug_node_1.asNativeElements;
exports.getDebugNode = debug_node_1.getDebugNode;
var testability_1 = require('./testability/testability');
exports.Testability = testability_1.Testability;
exports.TestabilityRegistry = testability_1.TestabilityRegistry;
exports.setTestabilityGetter = testability_1.setTestabilityGetter;
__export(require('./change_detection'));
__export(require('./platform_core_providers'));
var tokens_1 = require('./i18n/tokens');
exports.TRANSLATIONS = tokens_1.TRANSLATIONS;
exports.TRANSLATIONS_FORMAT = tokens_1.TRANSLATIONS_FORMAT;
exports.LOCALE_ID = tokens_1.LOCALE_ID;
var application_module_1 = require('./application_module');
exports.ApplicationModule = application_module_1.ApplicationModule;
var profile_1 = require('./profile/profile');
exports.wtfCreateScope = profile_1.wtfCreateScope;
exports.wtfLeave = profile_1.wtfLeave;
exports.wtfStartTimeRange = profile_1.wtfStartTimeRange;
exports.wtfEndTimeRange = profile_1.wtfEndTimeRange;
var type_1 = require('./type');
exports.Type = type_1.Type;
var async_1 = require('./facade/async');
exports.EventEmitter = async_1.EventEmitter;
var error_handler_1 = require('./error_handler');
exports.ErrorHandler = error_handler_1.ErrorHandler;
__export(require('./core_private_export'));
__export(require('./animation/metadata'));
var animation_transition_event_1 = require('./animation/animation_transition_event');
exports.AnimationTransitionEvent = animation_transition_event_1.AnimationTransitionEvent;
var animation_player_1 = require('./animation/animation_player');
exports.AnimationPlayer = animation_player_1.AnimationPlayer;
var animation_styles_1 = require('./animation/animation_styles');
exports.AnimationStyles = animation_styles_1.AnimationStyles;
var animation_keyframe_1 = require('./animation/animation_keyframe');
exports.AnimationKeyframe = animation_keyframe_1.AnimationKeyframe;
var security_1 = require('./security');
exports.Sanitizer = security_1.Sanitizer;
exports.SecurityContext = security_1.SecurityContext;
