// @angular/core/src/metadata.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
"use strict";
var di_1 = require('./metadata/di');
exports.ANALYZE_FOR_ENTRY_COMPONENTS = di_1.ANALYZE_FOR_ENTRY_COMPONENTS;
exports.Attribute = di_1.Attribute;
exports.ContentChild = di_1.ContentChild;
exports.ContentChildren = di_1.ContentChildren;
exports.Query = di_1.Query;
exports.ViewChild = di_1.ViewChild;
exports.ViewChildren = di_1.ViewChildren;
var directives_1 = require('./metadata/directives');
exports.Component = directives_1.Component;
exports.Directive = directives_1.Directive;
exports.HostBinding = directives_1.HostBinding;
exports.HostListener = directives_1.HostListener;
exports.Input = directives_1.Input;
exports.Output = directives_1.Output;
exports.Pipe = directives_1.Pipe;
var lifecycle_hooks_1 = require('./metadata/lifecycle_hooks');
exports.AfterContentChecked = lifecycle_hooks_1.AfterContentChecked;
exports.AfterContentInit = lifecycle_hooks_1.AfterContentInit;
exports.AfterViewChecked = lifecycle_hooks_1.AfterViewChecked;
exports.AfterViewInit = lifecycle_hooks_1.AfterViewInit;
exports.DoCheck = lifecycle_hooks_1.DoCheck;
exports.OnChanges = lifecycle_hooks_1.OnChanges;
exports.OnDestroy = lifecycle_hooks_1.OnDestroy;
exports.OnInit = lifecycle_hooks_1.OnInit;
var ng_module_1 = require('./metadata/ng_module');
exports.CUSTOM_ELEMENTS_SCHEMA = ng_module_1.CUSTOM_ELEMENTS_SCHEMA;
exports.NO_ERRORS_SCHEMA = ng_module_1.NO_ERRORS_SCHEMA;
exports.NgModule = ng_module_1.NgModule;
var view_1 = require('./metadata/view');
exports.ViewEncapsulation = view_1.ViewEncapsulation;
